building-an-api (#7)
Ada-Developers-Academy/core-unit-2: building-an-api/flask-hello-books.md

# Flask Hello Books

## Format

This lesson will be in the format of a walkthrough. It will use the [Hello Books API project repo](https://github.com/AdaGold/hello-books-api).

We will:

1. Clone our repo
1. Manage our dependencies
1. Define our first blueprint
1. Define our first endpoint, `/hello-world`
1. Witness the response of our first endpoint
1. Define our second endpoint, `/hello/JSON`
1. Witness the response of our second endpoint
1. Debug a third, broken endpoint

> We **highly recommend** reading or watching through this lesson first fully, before repeating on your own machine.
> 
> This lesson has a large amount of new information. It may be better to think about this lesson as a resource to be familiar with, rather than commit all new knowledge to memory at once.
> 
> ### !callout-danger
> 
> ## Read-through First, Repeat Second
> 
> We **highly suggest** reading or watching through this lesson fully, before repeating on your own machine.
> 
> ### !end-callout

Reviewer: I am not against giving this recommendation in other lessons but in this one it feels to me like the lesson is very much written for the student to work along as they read through the lesson. It is a series of experiments and actually isn't really (in my mind) that many new concepts. So I'd prefer removing these notes from this lesson in particular.

```suggestion
```

Reviewer 2: Initially, I thought more in line with the original text, that the student should go all the way through to have the next step in mind as they work through it on the seconds pass.

But now I think I agree with Jared.

I think a first pass would quickly turn into just skimming through to get to the end, and I think the majority of a video would go by me in this situation if I weren't following along.

Even if they misstep along the way and get confused, at worst they can delete everything and start over.

### Fork and Clone the Repo for Personal Experimentation

If you are interested in personal exploration of the project, fork this repo to your own account. This will let you have full control over the repo, and you can create and push commits over time.

## Clone and Explore

Clone the repo, `cd` into it, and open this project into your favorite text editor. We should see the following structure.

```
.
├── README.md
├── app
│   ├── __init__.py
│   └── routes.py
└── requirements.txt
```

The `app` directory will contain the bulk of our application code.

`app/routes.py` is the file we'll use the most in this lesson. It will be where we define our endpoints.

`app/__init__.py` is a file that contains a lot of configuration steps for our app.

## Manage Dependencies

Now, we need to create a virtual environment, activate it, and then install the dependencies in `requirements.txt`.

We can follow these steps in the Terminal:

```bash
$ python3 -m venv venv
$ source venv/bin/activate
(venv) $ pip install -r requirements.txt
```

### !callout-warning

## Common Debugging: Need to `$ pip install --upgrade pip`

> A common installation error may be solved by executing `$ pip install --upgrade pip`. Afterwards, attempt `$ pip install -r requirements.txt` again. _(Note: Written March 2021.)_

Author: This date supposed to shame us into double checking periodically! Taking out the dates from the others and leaving this top one.

In `routes.py`, type in the following code:

```python
from flask import Blueprint

hello_world_bp = Blueprint("hello_world_bp", __name__)
```

| <div style="min-width:200px;"> Piece of Code </div> | Notes                                                                                                                                                                                                     |
| --------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `hello_world_bp`                                    | The name for a local variable that will hold our Blueprint instance. Whenever we need to use this Blueprint to define a route, we'll use `hello_world_bp`.                                                |
| `Blueprint( ... )`                                  | The syntax for instantiating a `Blueprint`                                                                                                                                                                |

> | `"hello_world_bp"`                                  | The first argument is a string that will be used to name this `Blueprint` for debugging purposes. It's conventional to use the same name as the local variable holding the instance (`"hello_world_bp"`). |

Reviewer: ```suggestion
| `"hello_world_bp"`                                  | The first argument is a string that will be used to identify this `Blueprint` from the Flask server logs (in the terminal). It's conventional to use the same name as the local variable holding the instance (`"hello_world_bp"`). |
```

Reviewer 2: Building on Jared's suggestion (also incorporating the renamed string), since the var name and string _don't_ typically match, I'd provide some guidance here.
```suggestion
| `"hello_world"`                                     | The first argument is a string that will be used to identify this `Blueprint` from the Flask server logs (in the terminal). It's conventional to use a name related to the data being served or the functionality being provided. Here we are creating a "hello world" style endpoint, so "hello_world" is fine for our needs. |
```

> | `__name__`                                          | The second argument is the name of the Blueprint package. This is almost always `__name__`.                                                                                                               |

Reviewer 2: provide some context for `__name__` and what it's used for. I don't know what I should take from the argument being "the name of the Blueprint package".

```suggestion
| `__name__`                                          | The second argument is almost always the special Python variable `__name__`, which the blueprint uses to figure out certain aspects of routing.                                                                                                                                                                                |
```

### Registering a Blueprint

> Every time we instantiate a new `Blueprint`, Flask requires us to register it with `app`. This is a fancy way to say that we need to configure the app to recognize and use `hello_world_bp`.

Reviewer 2: more explicitly say what the app is doing with this

```suggestion
Every time we instantiate a new `Blueprint`, Flask requires us to register it with `app`. This is a fancy way to say that we need to tell the `app` that it should use the endpoints from `hello_world_bp` for its routing.
```

_Where_ we register the `Blueprint` depends on the project. For the Hello Books project specifically, the place to do this is in `app/__init__.py`, inside the `create_app()` function.

```python
from flask import Flask

def create_app(test_config=None):
    app = Flask(__name__)

    from .routes import hello_world_bp
    app.register_blueprint(hello_world_bp)

    return app
```

> | <div style="min-width:200px;"> Piece of Code </div> | Notes                                                                                                  |

Reviewer 2: widen to get a better word wrap
```suggestion
| <div style="min-width:240px;"> Piece of Code </div> | Notes                                                                                                  |
```

| `from .routes import hello_world_bp`                | We are importing `hello_world_bp` into this module so we may use it in the next line.                  |
| `app.register_blueprint(hello_world_bp)`            | We use `app`'s pre-defined function `register_blueprint()` to register the `hello_world_bp` Blueprint. |

If we needed to create and register more `Blueprint`s, we could follow this pattern and duplicate this code until we needed a better solution.

### Defining an Endpoint

Now that we have registered a Blueprint, which will help us organize our routes, we can create an endpoint in `app/routes.py`.

Recall that the responsibility of an endpoint is to:

1. Match the HTTP verb and request URL of an HTTP request
2. Form an HTTP response to send back to the client

Consider this explanation of syntax for a generic endpoint. An example of a specific endpoint is located later in the lesson.

```python
@blueprint_name.route("/endpoint/path/here", methods=["GET"])
def endpoint_name():
    my_beautiful_response_body = "Hello, World!"
    return my_beautiful_response_body
```

> | <div style="min-width:250px;"> Piece of Code </div> | Notes                                                                                                                                          |

Reviewer 2: widen to get a better word wrap
```suggestion
| <div style="min-width:290px;"> Piece of Code </div> | Notes                                                                                                                                                                                                                                       |
```

| --------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |

> | `@blueprint_name.route(...)`                        | This is a decorator for the endpoint. We use the defined Blueprint, and it's method `.route()`                                                 |

Reviewer: ```suggestion
| `@blueprint_name.route(...)`                        | This decorator transforms this function into an endpoint. We use the  `.route() instance method from our Blueprint instance.`                                                 |
```

Reviewer 2: again, building on Jared's suggestion (slight modification), but also, the code quote is out of place

```suggestion
| `@blueprint_name.route(...)`                        | This decorator transforms the function that follows into an endpoint. We use the `.route()` instance method from our Blueprint instance.                                                                                                    |
```

> | `"/endpoint/path/here"`                             | The request URL that must match                                                                                                                |
> | `methods=["GET"]`                                   | A list of HTTP methods that must match                                                                                                         |

Reviewer: I think we could offer a clearer explanation by explaining both arguments at once and how to combine together to define when this function will get called.
```suggestion
| `"/endpoint/path/here", methods=["GET"]`                             |  Together these arguments define what type of request  will be routed to this function. The first argument defines the path (or URL) of the request and the second argument defines a list of HTTP methods (or verbs) the request could have.                                                                                     |
```

> | `def endpoint_name():`                              | This function will execute whenever a request that matches the decorator is received. The name of this function can be made with any decision. |

Reviewer: I didn't really understand the 2nd sentence here so I took a guess at what it was trying to say and reworded it.
```suggestion
| `def endpoint_name():`                              | This function will execute whenever a request that matches the decorator is received. The function can be named whatever feels most appropriate. |
```

Reviewer 2: Building on Jared's suggestion
```suggestion
| `def endpoint_name():`                              | This function will execute whenever a request that matches the decorator is received. The function should be named using our typical practice of giving functions good, descriptive names. The name of the function has no effect on how requests are routed. |
```

| `my_beautiful_response_body = "Hello, World!"`      | We must define a response body to return. Here, we're using a local variable `my_beautiful_response_body` to hold a value                      |
| `return my_beautiful_response_body`                 | For each endpoint, we must _return_ the HTTP response.                                                                                         |

## Endpoint #1: `/hello-world`

Consider an endpoint definition that would:

- Use the Blueprint `hello_world_bp`
- Match the route `"/hello-world"`
- Match the HTTP method `["GET"]`
- Give a response `200 OK` with the HTTP body `"Hello, World!"`

> 

Reviewer 2: We explicitly say to use 200 OK, but don't say how to in the example and don't mention that it's the default. Suggest adding a callout about that.

```suggestion

### !callout-info

## `200 OK` Is the Default Response Status Code

We will see how to set explicit status codes in later lessons. If we don't set the status code ourselves, Flask will use `200 OK` by default.

### !end-callout

```

We may come up with something such as:

```python
from flask import Blueprint

> hello_world_bp = Blueprint("hello_world_bp", __name__)

Reviewer 2: matching the previous suggestion
```suggestion
hello_world_bp = Blueprint("hello_world", __name__)
```

@hello_world_bp.route("/hello-world", methods=["GET"])
def hello_world():

>     return "Hello, World!"

Reviewer: By making this the same as the example above, I think it will help students to more easily observe the meaningful differences between this code and the more generic code above.

```suggestion
    my_beautiful_response_body = "Hello, World!"
    return my_beautiful_response_body
```

```

### Manually Testing the Hello World Endpoint

_While the Flask server is running_, we can use Postman to send a `GET` request to `localhost:5000/hello-world`.

Hopefully, we see our HTTP response `200 OK` and `"Hello, World!"`!

![Screenshot of Postman after making a GET request to /hello-world with a 200 OK response](../assets/building-an-api/flask-hello-books_hello-world-postman.png)

Instead of using Postman, we can alternatively use our browser to make a `GET` request to `localhost:5000/hello-world`.

![Screenshot of a browser pointing to localhost:5000/hello-world with the text "Hello, World!"](../assets/building-an-api/flask-hello-books_hello-world-browser.png)

## Check the Server Logs

> Sending an HTTP request should come up as new details in the server logs.

Reviewer 2: Spell out what triggers the line, and what they should see.
```suggestion
Each time we send an HTTP request to our server, we should see a new line appear in the server log.
```

 * Debug mode: on
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
 * Debugger is active!
 * Debugger PIN: 231-804-583
127.0.0.1 - - [28/Mar/2021 17:08:19] "GET /hello-world HTTP/1.1" 200 -
```

### !callout-success

## Congratulations on Your First Endpoint!

Congratulations, you made your first endpoint in a Flask API! Congratulations to you and everyone around you! 🎉

### !end-callout

## Endpoint #2: `/hello/JSON`

Consider the code for a _second_ endpoint that would:

- Use the Blueprint `hello_world_bp`
- Match the route `"/hello/JSON"`
- Match the HTTP method `["GET"]`
- Give a response `200 OK`
- The HTTP response body should be the following JSON-like dictionary. Fill in the response body with your own details.

```json
{
  "name": "Ada Lovelace",
  "message": "Hello!",
  "hobbies": ["Fishing", "Swimming", "Watching Reality Shows"]
}
```

Your code could look similar to this:

```python
@hello_world_bp.route("/hello/JSON", methods=["GET"])
def hello_json():
    return {
        "name": "Ada Lovelace",
        "message": "Hello!",
        "hobbies": ["Fishing", "Swimming", "Watching Reality Shows"]
    }
```

### Manually Testing

> Confirm that you can send a request and get a response back with Postman.

Reviewer 2: revoicing for we
```suggestion
We can confirm that we can send a request and get a response back with Postman.
```

![Screenshot of Postman after making a GET request to /hello/JSON with a 200 OK response](../assets/building-an-api/flask-hello-books_hello-json-postman.png)

> Confirm that you can send a request and get a response back in the browser. This screenshot showcases a browser using a Google Chrome browser extension for viewing JSON.

Reviewer 2: revoicing for we
if the student doesn't have the plugin, it will look different, so mention that.

```suggestion
We can also confirm that we can send a request and get a response back in the browser. This screenshot showcases a Google Chrome browser extension for viewing JSON, so it's fine if the result doesn't look the same as this.
```

![Screenshot of a web browser after making a GET request to /hello/JSON with a JSON response body](../assets/building-an-api/flask-hello-books_hello-json-browser.png)

> Confirm that you can also see server log output.

Reviewer 2: revoicing for we
```suggestion
Finally, we can confirm that we can also see output in the server log.
```

![Screenshot of the terminal with the server log output showing that a GET request to /hello/JSON was made](../assets/building-an-api/flask-hello-books_hello-json-server.png)

### !callout-success

## Congratulations on Your Second Endpoint!

Congratulations on defining your _second_ endpoint! 🎉

### !end-callout

## Endpoint #3: Debugging a Broken Endpoint

> Place this third endpoint into `routes.py`. It is intentionally broken.

Reviewer 2: more invitational tone
```suggestion
Let's place this third endpoint into `routes.py`. It is intentionally broken.
```

```python
@hello_world_bp.route("/broken-endpoint-with-broken-server-code")
def broken_endpoint():
    response_body = {
        "name": "Ada Lovelace",
        "message": "Hello!",
        "hobbies": ["Fishing", "Swimming", "Watching Reality Shows"]
    }
    new_hobby = "Surfing"
    response_body["hobbies"] + new_hobby
    return response_body
```

When we make a `GET` request to `/broken-endpoint-with-broken-server-code`, we will get a `500 Internal Server Error` response.

![Screenshot of Postman after making a GET request to /broken-endpoint-with-broken-server-code with a 500 Internal Server Error response](../assets/building-an-api/flask-hello-books_hello-error-postman.png)

### Use Your Debugging Strategies

Use a combination of Postman, the browser, server logs, the Internet, and your peers. Fix this endpoint so it no longer sends back a `500` response, and instead sends back a `200` response.

Flask's error responses show up nicely formatted in the browser.

![Screenshot of the browser with Flask error messages](../assets/building-an-api/flask-hello-books_hello-error-browser.png)

> 

Reviewer 2: encourage checking the server log for the error (errors appear here even if the server isn't in debug mode)
also, add a callout for running the project in debug mode to get the error output shown in the screenshot
the callout is in the next comment (no suggestion because code)

```suggestion

We should also confirm that we can see the same error output in our server log.

```

Reviewer 2: ### !callout-info

## Flask Shows Helpful Error Messages in Debug Mode

If our browser shows only a `500 Internal Server Error` when we access this endpoint, we can restart Flask in debug mode. Running in debug mode will provide more details in the browser when an error is encountered in our server code. This information _could_ be used by malicious clients to learn about our server implementation, so we usually don't run deployed servers in debug mode.

<br />

To start Flask in debug mode, we should stop any current running Flask servers, then restart our server by running the following command.

```
FLASK_ENV=development flask run
```

### !end-callout

| `"hello_world_bp"`                                  | The first argument is a string that will be used to name this `Blueprint` for debugging purposes. It's conventional to use the same name as the local variable holding the instance (`"hello_world_bp"`). |
| `__name__`                                          | The second argument is the name of the Blueprint package. This is almost always `__name__`.                                                                                                               |

### Registering a Blueprint

Every time we instantiate a new `Blueprint`, Flask requires us to register it with `app`. This is a fancy way to say that we need to configure the app to recognize and use `hello_world_bp`.

_Where_ we register the `Blueprint` depends on the project. For the Hello Books project specifically, the place to do this is in `app/__init__.py`, inside the `create_app()` function.

```python
from flask import Flask

def create_app(test_config=None):
    app = Flask(__name__)

> 

Reviewer: I'd love for us to just add this without feeling the need to even explain it much because it makes development so much smoother and from my understanding there's no reason not to use it until deploying a project.
```suggestion
app.debug = True
```

Reviewer: This isn't successfully enabling debug mode for me with TaskList, so I think I'm missing something here.
@kaidamasaki do you have a quick guess about what might be missing?

Reviewer 2: We generally wouldn't hard code this into the app. Its usually controlled with either the FLASK_ENV=development or FLASK_DEBUG=1 environment variables.

If this _does_ get set explicitly, there should be a disclaimer that we are doing this out of expediency and that it generally would not be set this way in a deployed product.

Reviewer 2: VSCode also supports creating a flask launch configuration which sets the FLASK_ENV appropriately.
